os/fs/smartfs/smartfs_utils.c

```diff
@@ -1065,209 +1163,276 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
-#else
-			if ((entry->flags == SMARTFS_ERASEDSTATE_16BIT) || ((entry->flags &
-#endif
-					(SMARTFS_DIRENT_EMPTY | SMARTFS_DIRENT_ACTIVE)) == (~SMARTFS_ERASEDSTATE_16BIT & (SMARTFS_DIRENT_EMPTY | SMARTFS_DIRENT_ACTIVE)))) {
-				/* We found an empty entry.  Use it. */
+		while ((offset + entrysize) < readwrite.count) {
+			/* Check if this entry is available */
+			fdbg("\nEntry flags = %x\n", entry->flags);
+			if (IS_EMPTY_ENTRY(entry)) {
+				if (is_empty && entry->flags == SMARTFS_ERASEDSTATE_16BIT) {
+					fdbg("\nEmpty entry is foud\n");
+					*is_empty = TRUE;
+				}
 
-				found = TRUE;
-				break;
+				ret_entry->dsector = psector;
+				ret_entry->doffset = offset;
+				ret = OK;
+				goto errout;
 			}
 
-			/* Not available.  Skip to next entry */
-
+			/* Not an available entry, check the next one */
 			offset += entrysize;
 			entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
 		}
+		ret_entry->dsector = psector;
+		psector = nextsector;
+	}
 
-		/* If we found an entry, stop the search */
-
-		if (found) {
-			break;
-		}
-
-		/* If there are no more sectors, then we need to add one to make
-		 * room for the new entry.
-		 */
-
-		if (nextsector == SMARTFS_ERASEDSTATE_16BIT) {
-			/* Allocate a new sector and chain it to the last one */
-
-			ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
-			if (ret < 0) {
-				goto errout;
-			}
+errout:
+	fdbg("\nret = %d, invalidentry->dsector = %d, invalidentry->doffset = %d\n", ret, ret_entry->dsector, ret_entry->doffset);
+	if (ret != OK) {
+		ret = -ENOENT;
+	}
 
-			nextsector = (uint16_t)ret;
+	if (ret == OK) {
+		RESET_ENTRY_FLAGS(ret_entry);
+	}
 
-			/* Chain the next sector into this sector sector */
+	fdbg("\nret = %d, invalidentry->dsector = %d, invalidentry->doffset = %d\n", ret, ret_entry->dsector, ret_entry->doffset);
+	fdbg("\nfIND_FIRSTINVALIDENTRY returns invalidentry->flags = %x\n", ret_entry->flags);
+	return ret;
+}
 
-			chainheader->nextsector[0] = (uint8_t)(nextsector & 0x00FF);
-			chainheader->nextsector[1] = (uint8_t)(nextsector >> 8);
-			//*((uint16_t *)chainheader->nextsector) = nextsector;
-			readwrite.offset = offsetof(struct smartfs_chain_header_s, nextsector);
-			readwrite.count = sizeof(uint16_t);
-			readwrite.buffer = chainheader->nextsector;
-			ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
-			if (ret < 0) {
-				fdbg("Error chaining sector %d\n", nextsector);
-				goto errout;
-			}
+/****************************************************************************
+ * Name: smartfs_writeentry
+ *
+ * Description: Takes care of the operation of renaming an existing file
+ *              in a particular directory to a new name in the same dircetory
+ *              while minimizing the need to allocate new sectors for
+ *              writing the changed entry to MTD flash.
+ *
+ ****************************************************************************/
```

Comment:
Like a this function, it has many parameters and same type of params also there.(new_entry, ret_entry).
In this case you add comment regarding them.
* Input Parameters :
* fs
* new_entry
* blah
* lol
* rofl

Reply:
Yes, done.